# README

## diff1.pl [file]
count lines (total, blank, duplicated, unique) in one file

## diff2.pl [file1] [file2]
count unique lines in each file, and identical lines in two files

## ln.pl "path_to_dir/＊"
make soft link to files using wildcards

## name_space.pl [string1] [string2]
replace string1 using string2 in file name in current directory

## phptidy.php [php_file]
tidy php file for easy reading

## tsvtable.pl [tsv_file]
display tsv formated file as table-like output

## db/primer_format.pl [format-primer.txt]
batch add record into primer db

## db/add_sys_name.pl [file]
add systhesis name to primer

## venn_chart.pl [file1] [file2] [file2]
make venn chart between 2/3 file
modify Chart.pm for better font settting

## gi_ucsc_gene_transcript_seq.pl [gi.txt]
list gene exon-intron structure in excel format
print transcripts sequences in html, intron-exon highlighted

## kegg_rest_gene.pl [gene id / kegg pathway list]
retrieve kegg info

## split_by_list.pl [mirna_list.txt] [mirna_target.txt]
add kegg pathway enrichment annotation to mirna list

## primer_format.pl [primer_info.txt]
add primer info into db using batch script

## sfda_get_records.pl
从药监局网站获取国产器械的数据

## omim_gene_mutation.pl
extract gene mutation ALLELIC VARIANTS table from OMIM download data

## cv_xml_parse.pl
parse downloaded xml file in clinvar ftp site (ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar)

## known_gene_box.pl
draw a gene box in text file
turn off auto line wrap

## fas2iupac.pl [fasta file]
use clustal omega output(grouped fasta format) convert to iupac code sequence

## snp_summary_parse.pl
parse the summary formated result of ncbi snp result

## gi_clinvar_var.pl
get clinvar records and output mass array design file

## go_enrichment/
go enrichment analysis using R packages

## email-validation/
validate email address using php

## acc_ucsc_gene_transcript_seq.pl
get refseq exons to excel table

## get_refseq.pl
get refseq sequence of gi list
